### PR TITLE
graveler ref package implementation

### PR DESCRIPTION
**Pull request by nopcoder**

### CODE REVIEW

File: graveler/ref/branch_iterator_test.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+		DefaultBranchID:  "master",
+	}, graveler.Branch{}))
+
+	// prepare data
+	for _, b := range branches {
+		testutil.Must(t, r.SetBranch(ctx, "repo1", b, graveler.Branch{CommitID: "c1"}))
+	}
+
+	t.Run("listing all branches", func(t *testing.T) {
+		iter := ref.NewBranchIterator(ctx, db, "repo1", 3)
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:30]: 3 divides 6, so this doesn't test partial batching.  Can you add another branch on l.16, or change "3" to "4" or "5", please?

File: graveler/ref/branch_iterator_test.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.BranchID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
+
+		if diffs := deep.Equal(ids, []graveler.BranchID{"a", "aa", "b", "c", "d", "e", "master"}); diffs != nil {
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:41]: Does the interface guarantee branches are returned in sorted order?

File: graveler/ref/branch_iterator_test.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+			t.Fatalf("got wrong list of IDs: %v", diffs)
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:42]: nit:
```suggestion
			t.Errorf("got wrong list of IDs: %v", diffs)
```

File: graveler/ref/branch_iterator_test.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+		}
+	})
+
+	t.Run("listing branches from prefix", func(t *testing.T) {
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:46]: Not sure I understand the name (but the test is good and worthwhile!)

File: graveler/ref/branch_iterator_test.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
+
+		if diffs := deep.Equal(ids, []graveler.BranchID{"b", "c", "d", "e", "master"}); diffs != nil {
+			t.Fatalf("got wrong list of branch IDs: %v", diffs)
+		}
+
+		// now let's seek
+		iter.SeekGE("aa")
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:82]: This does not crash, which is a bug: future iterators _will_.  Let's make this iterator panic on user-after-Close?

File: graveler/ref/branch_iterator_test.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+		iter.SeekGE("b")
+		ids := make([]graveler.BranchID, 0)
+		for iter.Next() {
+			b := iter.Value()
+			ids = append(ids, b.BranchID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:75]:
```suggestion
```
(see also next comment)

File: graveler/ref/repository_iterator_test.go [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+		iter.SeekGE("b")
+		repoIds := make([]graveler.RepositoryID, 0)
+		for iter.Next() {
+			repo := iter.Value()
+			repoIds = append(repoIds, repo.RepositoryID)
+		}
+		if iter.Err() != nil {
+			t.Fatalf("unexpected error: %v", iter.Err())
+		}
+		iter.Close()
```
Comment (arielshaqed) [inline graveler/ref/repository_iterator_test.go:73]: Again use-after-Close works, because the PG-based implementation doesn't care.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Thanks!

Blocking is that use-after-Close works (and shouldn't), and that tests do that by mistake (which they shouldn't).  Everything else can be skipped or done later.

File: graveler/ref/branch_iterator.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	value        *graveler.BranchRecord
+	buf          []*graveler.BranchRecord
+
+	// blockValue is true when the iterator is created, or SeekGE() is called.
+	// A single Next() call turns it to false. When it's true, Value() returns nil.
+	blockValue bool
+
+	offset      string
+	fetchSize   int
+	shouldFetch bool
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply graveler/ref/branch_iterator.go:24]: can document it - 'shouldFetch' is true as an indicator for the initial state. on new iterator or after Seek, we shouldFetch is true, until we reach a point the last query doesn't fill our buffer.

File: graveler/ref/commit_walkter.go [hunk 9]
```diff
@@ -1,36 +1,35 @@
-package graveler
```
Comment (johnnyaug) [inline graveler/ref/commit_walkter.go:1]: typo in filename

File: graveler/ref/commit_walkter_test.go [hunk 10]
```diff
@@ -1,10 +1,11 @@
-package graveler_test
```
Comment (johnnyaug) [inline graveler/ref/commit_walkter_test.go:1]: typo in filename

**Review by johnnyaug — APPROVED** [review]
Cool!

File: graveler/ref/branch_iterator.go [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+	}
+}
+
+func (ri *BranchIterator) Next() bool {
+	ri.blockValue = false
+
+	// no buffer is initialized
+	if ri.buf == nil {
+		ri.fetch(true) // initial fetch
+	} else if len(ri.buf) == 0 {
```
[the remark this answers is not in the document]
Comment (nopcoder) [reply graveler/ref/branch_iterator.go]: I see your point - I'll rewrite this part using two booleans. Fetch called and fetch ended.

File: graveler/ref/branch_iterator_test.go (hunk 2 above)
Comment (nopcoder) [reply graveler/ref/branch_iterator_test.go:41]: For every iterator that we have SeekGE - the key is ordered

File: graveler/ref/branch_iterator_test.go (hunk 3 above)
Comment (nopcoder) [reply graveler/ref/branch_iterator_test.go:42]: why? just wonder as in this test there no other check after this one

File: graveler/ref/branch_iterator_test.go (hunk 4 above)
Comment (nopcoder) [reply graveler/ref/branch_iterator_test.go]: changed to "using prefix"

File: graveler/ref/branch_iterator.go [hunk 12]
```diff
@@ -8,22 +8,17 @@ import (
...
-	fetchSize   int
-	shouldFetch bool
-
-	err error
+	offset       string
+	fetchSize    int
+	err          error
+	fetchCalled  bool
+	fetchEnded   bool
+	closed       bool
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator.go:21]: Thanks, good names!  Now I see and understand: _it's a state machine_!  Not in this PR, but I would like to try refactor it to have one explicit state.  Think it's worth a shot?

File: graveler/ref/branch_iterator.go [hunk 13]
```diff
@@ -38,43 +33,44 @@ func NewBranchIterator(ctx context.Context, db db.Database, repositoryID gravele
...
+	if len(ri.buf) == 0 {
+		return false
+	}
 	ri.value = ri.buf[0]
 	ri.offset = string(ri.value.BranchID)
 	if len(ri.buf) > 1 {
 		ri.buf = ri.buf[1:]
 	} else {
-		ri.buf = make([]*graveler.BranchRecord, 0)
+		ri.buf = ri.buf[:0]
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator.go:58]: # :+1:

File: graveler/ref/iterator.go [hunk 14]
```diff
@@ -1,10 +1,14 @@
 package ref
 
+import "errors"
+
 const (
 	iteratorOffsetGE = ">="
 	iteratorOffsetGT = ">"
 )
 
+var ErrIteratorClosed = errors.New("iterator already closed")
```
Comment (arielshaqed) [inline graveler/ref/iterator.go:10]: All of these should be the same error.

Also: what is The Go Dan to keep these changes sync'ed?  I'm reviewing these by reading through them, but this is a *horrible* way to do it.  Surely we're not the first people to run into this...

File: graveler/ref/repository_iterator_test.go [hunk 15]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/treeverse/lakefs/testutil"
 )
 
-func TestPGRepositoryIterator(t *testing.T) {
+func TestRepositoryIterator(t *testing.T) {
 	r, db := testRefManagerWithDB(t)
 	repos := []graveler.RepositoryID{"a", "aa", "b", "c", "e", "d"}
```
Comment (arielshaqed) [inline graveler/ref/repository_iterator_test.go:16]: 6 is still divisible by 3, I'd rather we also tested a partial fetch.

File: graveler/ref/branch_iterator_test.go [hunk 16]
```diff
@@ -11,7 +11,7 @@ import (
 	"github.com/treeverse/lakefs/testutil"
 )
 
-func TestPGBranchIterator(t *testing.T) {
+func TestBranchIterator(t *testing.T) {
 	r, db := testRefManagerWithDB(t)
 	branches := []graveler.BranchID{"a", "aa", "b", "c", "e", "d"}
```
Comment (arielshaqed) [inline graveler/ref/branch_iterator_test.go:16]: 6 divisible by 3; I think this is unchanged from the previous review where it means that we end up not testing partial fetches (we should do both).  Or, perhaps run the "listing all branches" test in a loop for _all_ batch sizes 1..7?

**Review by arielshaqed — APPROVED** [review]
Great, thanks!
Some comments inside, nothing blocking.

File: graveler/ref/branch_iterator_test.go (hunk 16 above)
Comment (nopcoder) [reply graveler/ref/branch_iterator_test.go:16]: There are 6 branches here + master = so we have 7. I'll make sure tag and repo will have the same.

File: graveler/ref/branch_iterator_test.go (hunk 1 above)
Comment (nopcoder) [reply graveler/ref/branch_iterator_test.go:30]: keeping branch iterator working on 7 items.

File: graveler/ref/commit_walkter.go (hunk 9 above)
Comment (nopcoder) [reply graveler/ref/commit_walkter.go:1]: Thanks!

File: graveler/ref/repository_iterator_test.go (hunk 15 above)
Comment (nopcoder) [reply graveler/ref/repository_iterator_test.go:16]: fixed

File: graveler/ref/iterator.go (hunk 14 above)
Comment (nopcoder) [reply graveler/ref/iterator.go:10]: https://github.com/googleapis/google-api-go-client/blob/v0.36.0/iterator/iterator.go

can modify the code not to panic - just report an error. can condition the SeekGE not to reset the error just for this case / state.

What do you think?

Comment (arielshaqed) [reply graveler/ref/iterator.go:19]: Yeah.  Probably not in this PR, though :-)

File: graveler/ref/branch_iterator_test.go (hunk 16 above)
Comment (arielshaqed) [reply graveler/ref/branch_iterator_test.go:16]: Ouch!  A rich corporation once said, "[math is hard](https://whyplay.mattel.com/the-barbie-vlog-empowers-girls-with-words/)" -- and they're still not sorry!

**Review by arielshaqed — APPROVED** [review]
Thanks!

File: graveler/ref/iterator.go (hunk 14 above)
Comment (nopcoder) [reply graveler/ref/iterator.go:19]: Don't stop - we will get it better next round

**Pull request merged**